Mathlib/Topology/Algebra/ProperAction.lean

```diff
@@ -0,0 +1,342 @@
+  isProperMap_smul_pair' := by
+    have := isProperMap_of_closedEmbedding f_clemb
+    have : IsProperMap (Prod.map f (fun x : X ↦ x)) := IsProperMap.prod_map this isProperMap_id
+    have : (fun hx : H × X ↦ (hx.1 • hx.2, hx.2)) = (fun hx ↦ (f hx.1 • hx.2, hx.2)) := by
+      simp [f_compat]
+    rw [this]
+    change IsProperMap ((fun gx : G × X ↦ (gx.1 • gx.2, gx.2)) ∘ (Prod.map f (fun x ↦ x)))
+    apply IsProperMap.comp
+    assumption
+    exact isProperMap_smul_pair G X
+
+/-- If `H` is a closed subgroup of `G` and `G` acts properly on X then so does `H`. -/
+@[to_additive "If `H` is a closed subgroup of `G` and `G` acts properly on X then so does `H`."]
+instance {H : Subgroup G} [ProperSMul G X] [H_closed : IsClosed (H : Set G)] : ProperSMul H X where
+  isProperMap_smul_pair' := by
+    have : IsProperMap (fun hx : H × X ↦ ((hx.1, hx.2) : G × X)) := by
+      change IsProperMap (Prod.map ((↑) : H → G) (fun x ↦ x))
+      exact IsProperMap.prod_map (isProperMap_subtype_val_of_closed H_closed) isProperMap_id
+    have : IsProperMap (fun hx : H × X ↦ (hx.1 • hx.2, hx.2)) := by
+      change IsProperMap ((fun gx ↦ (gx.1 • gx.2, gx.2)) ∘
+        (fun hx : H × X ↦ ((hx.1, hx.2) : G × X)))
+      exact this.comp (isProperMap_smul_pair G X)
+    assumption
+
+/-
+Some suggestions of things to prove,
+taken from Kapovich
+-/
+
+/-
+If `X` is T2 and first countable,
+then the naive definition
+of proper action is equivalent to the good definition
+(Kapovich Th 11)
+-/
+theorem naiveProper_iff_ProperSMul_T2_FirstCountable
+    [T2Space X] [FirstCountableTopology X] :
+    ProperlyDiscontinuousSMul G X ↔ ProperSMul G X
+    := by sorry
+
+lemma tendsTo_comp_continuous
+    {lx: Filter X} {f : X → Y} {g : Y → Z} {y : Y}
+    (H : Tendsto f lx (𝓝 y)) (hg: Continuous g) :
+    Tendsto (g ∘ f) lx (𝓝 (g y)) := by
+  apply Filter.Tendsto.comp _ H
+  exact hg.tendsto y
+
+/-
+If `X` and `Y` are T2 and first countable,
+then the naive definition
+of proper map is equivalent to the good definition
+-/
+theorem properMap_of_naiveProper_T2_FirstCountable
+    [T2Space X] [FirstCountableTopology X]
+    [T2Space Y] [FirstCountableTopology Y]
+    (f : X → Y) (hcont: Continuous f):
+    (∀ (K : Set Y), (IsCompact K → IsCompact (f ⁻¹' K)))
+    → IsProperMap f
+    := by
+  intro h
+  -- a map is proper iff it is closed and the fibers are compacts
+  rw [isProperMap_iff_isClosedMap_and_compact_fibers]
+  refine ⟨(by assumption), ?_, ?_ ⟩
+  · rw [IsClosedMap]
+    intro U closed_U
+    -- in a first countable space, a set is closed iff sequentially closed
+    apply IsSeqClosed.isClosed
+    -- introduce a converging sequence in the image f(U)
+    intro seq y seq_in_image seq_conv
+    -- pullback this sequence to a sequence in U
+    choose s' hs using seq_in_image
+    have s'_in_U := fun n ↦ (hs n).1
+    have seq_factor : seq = f ∘ s' := by
+      ext n
+      simp
+      rw [<-(hs n).2]
```

Comment:
Here the linter complains about a "non-terminal simp". I don't remember how much we talked about it during the week, but here's what it means:

Since the behaviour of `simp` is determined by the set of lemmas which are tagged `simp`, it can change very often, sometimes due to seemingly unrelated changes. Most of the time these changes make `simp` more powerful, which means that if `simp` used to close the goal, it will probably still be able to do so after the change. This means that using `simp` is perfectly fine *as long as it closes the current goal*. If that is not the case, then `simp` may be able to do more progress by becoming more powerful, which could break the rest of the proof.

The typical fix for this is to use `simp?` and use the suggested `simp only`. But in this case all it does is unfold composition, which you can avoid entirely by doing
```suggestion
      ext n
      rw [<-(hs n).2]
      rfl
```
or even
```suggestion
      ext n
      exact (hs n).2.symm
```